### PR TITLE
Add Missing Test Function Related to Plotly.

### PR DESCRIPTION
**Pull request by MasahitoKumada**
## Motivation

 
Related to https://github.com/optuna/optuna/issues/3226

## Description of the changes

Add missing test functions in tests/visualization_tests/test_parallel_coordinate.py.

### CODE REVIEW

File: tests/visualization_tests/test_parallel_coordinate.py [hunk 1]
```diff
@@ -179,6 +218,41 @@ def test_plot_parallel_coordinate_log_params() -> None:
...
+            distributions={
+                "category_a": CategoricalDistribution(("preferred", "opt")),
+                "param_b": LogUniformDistribution(1, 1000),
+            },
+        )
+    )
+
+    # both hyperparameters contain unique values
+    figure = plot_parallel_coordinate(study_categorical_params)
+    assert len(figure.get_lines()) == 0
```
Comment (nzw0301) [inline tests/visualization_tests/test_parallel_coordinate.py:238]:
```suggestion
assert len(figure.data[0]["dimensions"]) == 3
    assert figure.data[0]["dimensions"][0]["label"] == "Objective Value"
    assert figure.data[0]["dimensions"][0]["range"] == (0.0, 0.0)
    assert figure.data[0]["dimensions"][0]["values"] == (0.0,)
    assert figure.data[0]["dimensions"][1]["label"] == "category_a"
    assert figure.data[0]["dimensions"][1]["range"] == (0, 0)
    assert figure.data[0]["dimensions"][1]["values"] == (0.0,)
    assert figure.data[0]["dimensions"][1]["ticktext"] == ("preferred",)
    assert figure.data[0]["dimensions"][2]["label"] == "param_b"
    assert figure.data[0]["dimensions"][2]["range"] == (math.log10(30), math.log10(30))
    assert figure.data[0]["dimensions"][2]["values"] == (math.log10(30),)
    assert figure.data[0]["dimensions"][2]["ticktext"] == ("30",)
    assert figure.data[0]["dimensions"][2]["tickvals"] == (math.log10(30),)
```

File: tests/visualization_tests/test_parallel_coordinate.py [hunk 2]
```diff
@@ -179,6 +218,41 @@ def test_plot_parallel_coordinate_log_params() -> None:
...
+            distributions={
+                "category_a": CategoricalDistribution(("preferred", "opt")),
+                "param_b": LogUniformDistribution(1, 1000),
+            },
+        )
+    )
+
+    # still "category_a" contains unique suggested value during the optimization.
+    figure = plot_parallel_coordinate(study_categorical_params)
+    assert len(figure.get_lines()) == 0
```
Comment (nzw0301) [inline tests/visualization_tests/test_parallel_coordinate.py:253]:
```suggestion
    assert len(figure.data[0]["dimensions"]) == 3
    assert figure.data[0]["dimensions"][1]["label"] == "category_a"
    assert figure.data[0]["dimensions"][1]["range"] == (0, 0)
    assert figure.data[0]["dimensions"][1]["values"] == (0.0, 0.0)
    assert figure.data[0]["dimensions"][1]["ticktext"] == ("preferred",)

```

**Review by nzw0301 — CHANGES_REQUESTED** [review]
I left two suggestions for `test_plot_parallel_coordinate_unique_hyper_param`

1. not using matplotlib syntax
2. Add more detailed tests by following other test functions in this file

File: tests/visualization_tests/test_parallel_coordinate.py (hunk 1 above)
Comment (MasahitoKumada) [reply tests/visualization_tests/test_parallel_coordinate.py:238]: Thank you for your suggestion on the revision details. I will reflect them.

File: tests/visualization_tests/test_parallel_coordinate.py (hunk 2 above)
Comment (MasahitoKumada) [reply tests/visualization_tests/test_parallel_coordinate.py:265]: Thank you for your suggestion on the revision details. I will reflect them.

File: tests/visualization_tests/test_parallel_coordinate.py [hunk 3]
```diff
@@ -129,6 +129,93 @@ def test_plot_parallel_coordinate_categorical_params() -> None:
...
+    assert figure.data[0]["dimensions"][1]["label"] == "category_a"
+    assert figure.data[0]["dimensions"][1]["range"] == (0, 0)
+    assert figure.data[0]["dimensions"][1]["values"] == (0.0,)
+    assert figure.data[0]["dimensions"][1]["ticktext"] == ("2",)
+    assert figure.data[0]["dimensions"][2]["label"] == "category_b"
+    assert figure.data[0]["dimensions"][2]["range"] == (0, 0)
+    assert figure.data[0]["dimensions"][2]["values"] == (0.0,)
+    assert figure.data[0]["dimensions"][2]["ticktext"] == ("20",)
+    assert figure.data[0]["dimensions"][2]["tickvals"] == (0,)
+
```
Comment (nzw0301) [inline tests/visualization_tests/test_parallel_coordinate.py:161]:
```suggestion
```

File: tests/visualization_tests/test_parallel_coordinate.py [hunk 4]
```diff
@@ -129,6 +129,93 @@ def test_plot_parallel_coordinate_categorical_params() -> None:
...
+    assert figure.data[0]["dimensions"][1]["values"] == (0.0,)
+    assert figure.data[0]["dimensions"][1]["ticktext"] == ("1",)
+    assert figure.data[0]["dimensions"][2]["label"] == "category_b"
+    assert figure.data[0]["dimensions"][2]["range"] == (0, 0)
+    assert figure.data[0]["dimensions"][2]["values"] == (0.0,)
+    assert figure.data[0]["dimensions"][2]["ticktext"] == ("30",)
+    assert figure.data[0]["dimensions"][2]["tickvals"] == (0,)
+
+
+
```
Comment (nzw0301) [inline tests/visualization_tests/test_parallel_coordinate.py:190]:
```suggestion
```

File: tests/visualization_tests/test_parallel_coordinate.py [hunk 5]
```diff
@@ -129,6 +129,93 @@ def test_plot_parallel_coordinate_categorical_params() -> None:
...
+    assert figure.data[0]["dimensions"][0]["values"] == (0.0,)
+    assert figure.data[0]["dimensions"][1]["label"] == "category_a"
+    assert figure.data[0]["dimensions"][1]["range"] == (0, 0)
+    assert figure.data[0]["dimensions"][1]["values"] == (0.0,)
+    assert figure.data[0]["dimensions"][1]["ticktext"] == ("2",)
+    assert figure.data[0]["dimensions"][2]["label"] == "category_b"
+    assert figure.data[0]["dimensions"][2]["range"] == (0, 0)
+    assert figure.data[0]["dimensions"][2]["values"] == (0.0,)
+    assert figure.data[0]["dimensions"][2]["ticktext"] == ("10",)
+    assert figure.data[0]["dimensions"][2]["tickvals"] == (0,)
```
Comment (nzw0301) [inline tests/visualization_tests/test_parallel_coordinate.py:216]:
```suggestion
    # Trials are sorted by using param_a and param_b.
    figure = plot_parallel_coordinate(study_categorical_params)
    assert len(figure.data[0]["dimensions"]) == 3
    assert figure.data[0]["dimensions"][0]["label"] == "Objective Value"
    assert figure.data[0]["dimensions"][0]["range"] == (0.0, 2.0)
    assert figure.data[0]["dimensions"][0]["values"] == (1.0, 2.0, 0.0)
    assert figure.data[0]["dimensions"][1]["label"] == "category_a"
    assert figure.data[0]["dimensions"][1]["range"] == (0, 1)
    assert figure.data[0]["dimensions"][1]["values"] == (0, 1, 1)
    assert figure.data[0]["dimensions"][1]["ticktext"] == (1, 2)
    assert figure.data[0]["dimensions"][1]["tickvals"] == (0, 1)
    assert figure.data[0]["dimensions"][2]["label"] == "category_b"
    assert figure.data[0]["dimensions"][2]["range"] == (0, 2)
    assert figure.data[0]["dimensions"][2]["values"] == (2, 0, 1)
    assert figure.data[0]["dimensions"][2]["ticktext"] == (10, 20, 30)
    assert figure.data[0]["dimensions"][2]["tickvals"] == (0, 1, 2)
```

File: tests/visualization_tests/test_parallel_coordinate.py [hunk 6]
```diff
@@ -179,6 +266,58 @@ def test_plot_parallel_coordinate_log_params() -> None:
...
+        )
+    )
+
+    # still "category_a" contains unique suggested value during the optimization.
+    figure = plot_parallel_coordinate(study_categorical_params)
+    assert len(figure.data[0]["dimensions"]) == 3
+    assert figure.data[0]["dimensions"][1]["label"] == "category_a"
+    assert figure.data[0]["dimensions"][1]["range"] == (0, 0)
+    assert figure.data[0]["dimensions"][1]["values"] == (0.0, 0.0)
+    assert figure.data[0]["dimensions"][1]["ticktext"] == ("preferred",)
```
Comment (nzw0301) [inline tests/visualization_tests/test_parallel_coordinate.py:317]:
```suggestion
    assert figure.data[0]["dimensions"][1]["ticktext"] == ("preferred",)
    assert figure.data[0]["dimensions"][1]["tickvals"] == (0,)    
```

File: tests/visualization_tests/test_parallel_coordinate.py [hunk 7]
```diff
@@ -179,6 +266,58 @@ def test_plot_parallel_coordinate_log_params() -> None:
...
+    # both hyperparameters contain unique values
+    figure = plot_parallel_coordinate(study_categorical_params)
+    assert len(figure.data[0]["dimensions"]) == 3
+    assert figure.data[0]["dimensions"][0]["label"] == "Objective Value"
+    assert figure.data[0]["dimensions"][0]["range"] == (0.0, 0.0)
+    assert figure.data[0]["dimensions"][0]["values"] == (0.0,)
+    assert figure.data[0]["dimensions"][1]["label"] == "category_a"
+    assert figure.data[0]["dimensions"][1]["range"] == (0, 0)
+    assert figure.data[0]["dimensions"][1]["values"] == (0.0,)
+    assert figure.data[0]["dimensions"][1]["ticktext"] == ("preferred",)
```
Comment (nzw0301) [inline tests/visualization_tests/test_parallel_coordinate.py:293]:
```suggestion
    assert figure.data[0]["dimensions"][1]["ticktext"] == ("preferred",)
    assert figure.data[0]["dimensions"][1]["tickvals"] == (0,)    
```

**Review by nzw0301 — CHANGES_REQUESTED** [review]
I would like to suggest replacing the test assertions of `test_plot_parallel_coordinate_categorical_numeric_params` with more minimal assertions. In addition, I suggested adding missing assertions to `test_plot_parallel_coordinate_unique_hyper_param`.

File: tests/visualization_tests/test_parallel_coordinate.py (hunk 3 above)
Comment (MasahitoKumada) [reply tests/visualization_tests/test_parallel_coordinate.py:161]: Reflect the correction.

File: tests/visualization_tests/test_parallel_coordinate.py [hunk 8]
```diff
@@ -179,6 +231,60 @@ def test_plot_parallel_coordinate_log_params() -> None:
...
+    assert figure.data[0]["dimensions"][1]["tickvals"] == (0,)
```
Comment (nzw0301) [inline tests/visualization_tests/test_parallel_coordinate.py:259]:
```suggestion
    assert figure.data[0]["dimensions"][1]["tickvals"] == (0,)
```

File: tests/visualization_tests/test_parallel_coordinate.py [hunk 9]
```diff
@@ -179,6 +231,60 @@ def test_plot_parallel_coordinate_log_params() -> None:
...
+    )
+
+    # still "category_a" contains unique suggested value during the optimization.
+    figure = plot_parallel_coordinate(study_categorical_params)
+    assert len(figure.data[0]["dimensions"]) == 3
+    assert figure.data[0]["dimensions"][1]["label"] == "category_a"
+    assert figure.data[0]["dimensions"][1]["range"] == (0, 0)
+    assert figure.data[0]["dimensions"][1]["values"] == (0.0, 0.0)
+    assert figure.data[0]["dimensions"][1]["ticktext"] == ("preferred",)
+    assert figure.data[0]["dimensions"][1]["tickvals"] == (0,)
```
Comment (nzw0301) [inline tests/visualization_tests/test_parallel_coordinate.py:284]:
```suggestion
    assert figure.data[0]["dimensions"][1]["tickvals"] == (0,)
```

File: tests/visualization_tests/test_parallel_coordinate.py [hunk 10]
```diff
@@ -179,6 +231,60 @@ def test_plot_parallel_coordinate_log_params() -> None:
...
+
+    # still "category_a" contains unique suggested value during the optimization.
+    figure = plot_parallel_coordinate(study_categorical_params)
+    assert len(figure.data[0]["dimensions"]) == 3
+    assert figure.data[0]["dimensions"][1]["label"] == "category_a"
+    assert figure.data[0]["dimensions"][1]["range"] == (0, 0)
+    assert figure.data[0]["dimensions"][1]["values"] == (0.0, 0.0)
+    assert figure.data[0]["dimensions"][1]["ticktext"] == ("preferred",)
+    assert figure.data[0]["dimensions"][1]["tickvals"] == (0,)    
+
```
Comment (nzw0301) [inline tests/visualization_tests/test_parallel_coordinate.py:285]:
```suggestion
```

**Review by nzw0301 — CHANGES_REQUESTED** [review]
For `black`, could you update the PR by following suggestions?

File: tests/visualization_tests/test_parallel_coordinate.py [hunk 11]
```diff
@@ -179,6 +231,60 @@ def test_plot_parallel_coordinate_log_params() -> None:
...
+            value=0.0,
+            params={"category_a": "preferred", "param_b": 30},
+            distributions={
+                "category_a": CategoricalDistribution(("preferred", "opt")),
+                "param_b": LogUniformDistribution(1, 1000),
+            },
+        )
+    )
+
+    # both hyperparameters contain unique values
```
Comment (nzw0301) [inline tests/visualization_tests/test_parallel_coordinate.py:249]: Sorry, one more comment on here
```suggestion
    # Both hyperparameters contain unique values.
```

File: tests/visualization_tests/test_parallel_coordinate.py [hunk 12]
```diff
@@ -179,6 +231,59 @@ def test_plot_parallel_coordinate_log_params() -> None:
...
+            value=2.0,
+            params={"category_a": "preferred", "param_b": 20},
+            distributions={
+                "category_a": CategoricalDistribution(("preferred", "opt")),
+                "param_b": LogUniformDistribution(1, 1000),
+            },
+        )
+    )
+
+    # still "category_a" contains unique suggested value during the optimization.
```
Comment (nzw0301) [inline tests/visualization_tests/test_parallel_coordinate.py:277]:
```suggestion
    # Still "category_a" contains unique suggested value during the optimization.
```

**Review by nzw0301 — APPROVED** [review]

File: tests/visualization_tests/test_parallel_coordinate.py [hunk 13]
```diff
@@ -129,6 +129,58 @@ def test_plot_parallel_coordinate_categorical_params() -> None:
...
+            value=2.0,
+            params={"category_a": 2, "category_b": 10},
+            distributions={
+                "category_a": CategoricalDistribution((1, 2)),
+                "category_b": CategoricalDistribution((10, 20, 30)),
+            },
+        )
+    )
+
+    # Trials are sorted by using param_a and param_b.
```
Comment (toshihikoyanase) [inline tests/visualization_tests/test_parallel_coordinate.py:166]: Could you add the expected order of trial in concrete? 
```suggestion
    # Trials are sorted by using param_a and param_b, i.e., trial#1, trial#2, and trial#0.
```

**Review by toshihikoyanase — COMMENTED** [review]
Thank you for your PR. The change looks great to me, but I added a small comment.

**Review by toshihikoyanase — APPROVED** [review]
Thank you for your update! LGTM.

**Comment by nzw0301** [comment]
Let me rename the PR title with an informative one for the forthcoming release note.

**Pull request merged**